<a href="https://colab.research.google.com/github/foweiss/hello-world/blob/master/Final_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Performance of Simple Neural Network (NN) vs Convolutional Neural Network (CNN) in handwritten digit recognition using MNIST Dataset

1- The simple neural network stucture is:


*   Input layer has 784 inputs (28*28)
*   2 fully connected hidden layers each having 384 neurons
*   Output layer having 10 neurons that correspond to the 10 classes





In [0]:
!pip install --upgrade tensorflow

In [1]:
import tensorflow as tf
tf.__version__

'2.0.0'

In [0]:
import numpy as np
from tensorflow import keras
from keras.datasets import mnist
from keras import backend as K

In [0]:
# Import data split as train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Number of classes (0, 1, 2 ,... ,9)
num_classes = 10

# Convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

# Reshape data for input layer of NN
x_train_r = x_train.reshape(60000, 784)
x_test_r = x_test.reshape(10000, 784)
# Convert data to float type
x_train_r = x_train_r.astype('float32')
x_test_r = x_test_r.astype('float32')
# Normalize data
x_train_r /= 255
x_test_r /= 255

In [21]:
# Simple NN
model_nn = tf.keras.models.Sequential()
model_nn.add(tf.keras.layers.Dense(384, activation="relu",  input_shape=(784,)))
model_nn.add(tf.keras.layers.Dropout(0.2))
model_nn.add(tf.keras.layers.Dense(384, activation="relu"))
model_nn.add(tf.keras.layers.Dropout(0.2))
model_nn.add(tf.keras.layers.Dense(num_classes, activation="softmax"))


model_nn.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 384)               301440    
_________________________________________________________________
dropout (Dropout)            (None, 384)               0         
_________________________________________________________________
dense_13 (Dense)             (None, 384)               147840    
_________________________________________________________________
dropout_1 (Dropout)          (None, 384)               0         
_________________________________________________________________
dense_14 (Dense)             (None, 10)                3850      
Total params: 453,130
Trainable params: 453,130
Non-trainable params: 0
_________________________________________________________________


In [22]:
model_nn.compile(loss='categorical_crossentropy',
                 optimizer='sgd',
                 metrics=['accuracy'])
history_nn = model_nn.fit(x_train_r, y_train,
                    epochs=10,
                    verbose=1,
                    validation_data=(x_test_r, y_test))
score = model_nn.evaluate(x_test_r, y_test, verbose=0)
print('Test loss_nn:', score[0])
print('Test accuracy_nn:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 13s 215us/sample - loss: 0.6737 - accuracy: 0.8091 - val_loss: 0.3095 - val_accuracy: 0.9119
Epoch 2/10
60000/60000 [==============================] - 12s 206us/sample - loss: 0.3287 - accuracy: 0.9035 - val_loss: 0.2398 - val_accuracy: 0.9309
Epoch 3/10
60000/60000 [==============================] - 12s 206us/sample - loss: 0.2662 - accuracy: 0.9230 - val_loss: 0.2021 - val_accuracy: 0.9422
Epoch 4/10
60000/60000 [==============================] - 13s 209us/sample - loss: 0.2226 - accuracy: 0.9349 - val_loss: 0.1702 - val_accuracy: 0.9493
Epoch 5/10
60000/60000 [==============================] - 13s 212us/sample - loss: 0.1968 - accuracy: 0.9423 - val_loss: 0.1520 - val_accuracy: 0.9554
Epoch 6/10
60000/60000 [==============================] - 13s 210us/sample - loss: 0.1765 - accuracy: 0.9477 - val_loss: 0.1381 - val_accuracy: 0.9590
Epoch 7/10
60000/60000 [====================

2- The CNN structure is:


*   2D Convolutional layer with 32 filters
*   2D Convolutional layer with 64 filters
*   MaxPooling layer of size 2x2
*   Flatten layer to reshape array
*   Fully connected layer of 128 neurons
*   Output layer having 10 neurons that correspond to the 10 classes



In [40]:
# input image dimensions
img_rows, img_cols = 28, 28

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

# Convert data to float type
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
# Normalize data
x_train /= 255
x_test /= 255

# CNN layers
model_cnn = tf.keras.models.Sequential()
model_cnn.add(tf.keras.layers.Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model_cnn.add(tf.keras.layers.Conv2D(64, (3, 3), activation='relu'))
model_cnn.add(tf.keras.layers.MaxPooling2D(pool_size=(2, 2)))
model_cnn.add(tf.keras.layers.Dropout(0.2))
model_cnn.add(tf.keras.layers.Flatten())
model_cnn.add(tf.keras.layers.Dense(128, activation='relu'))
model_cnn.add(tf.keras.layers.Dropout(0.2))
model_cnn.add(tf.keras.layers.Dense(num_classes, activation='softmax'))

model_cnn.summary()

Model: "sequential_15"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_20 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_21 (Conv2D)           (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 12, 12, 64)        0         
_________________________________________________________________
dropout_22 (Dropout)         (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_13 (Flatten)         (None, 9216)              0         
_________________________________________________________________
dense_35 (Dense)             (None, 128)               1179776   
_________________________________________________________________
dropout_23 (Dropout)         (None, 128)             

In [41]:
# Compile and run CNN model
model_cnn.compile(loss='categorical_crossentropy',
                 optimizer='sgd',
                 metrics=['accuracy'])

model_cnn.fit(x_train, y_train,    
          epochs=10,
          verbose=1,
          validation_data=(x_test, y_test))
score = model_cnn.evaluate(x_test, y_test, verbose=0)
print('Test loss_CNN:', score[0])
print('Test accuracy_CNN:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 189s 3ms/sample - loss: 2.3016 - accuracy: 0.1116 - val_loss: 2.3011 - val_accuracy: 0.1135
Epoch 2/10
60000/60000 [==============================] - 188s 3ms/sample - loss: 2.3013 - accuracy: 0.1124 - val_loss: 2.3009 - val_accuracy: 0.1135
Epoch 3/10
60000/60000 [==============================] - 188s 3ms/sample - loss: 2.3013 - accuracy: 0.1124 - val_loss: 2.3010 - val_accuracy: 0.1135
Epoch 4/10
60000/60000 [==============================] - 188s 3ms/sample - loss: 2.3013 - accuracy: 0.1124 - val_loss: 2.3011 - val_accuracy: 0.1135
Epoch 5/10
60000/60000 [==============================] - 189s 3ms/sample - loss: 2.3013 - accuracy: 0.1124 - val_loss: 2.3010 - val_accuracy: 0.1135
Epoch 6/10
60000/60000 [==============================] - 189s 3ms/sample - loss: 2.3014 - accuracy: 0.1124 - val_loss: 2.3010 - val_accuracy: 0.1135
Epoch 7/10
60000/60000 [==========================

Results

For this example, the CNN model did not work properly, it took 32 minutes to train and there are close to 1.2 million trainable parameters.
On the other hand, the simple NN model took less than 3 minutes to train, had a test accuracy of 96.87% and had 453,130 trainable parameters (about one third of the CNN).



Conclusion

For the MNIST handwritten digit recognition, a simple neural network gives acceptable results.
